crates/rspack_core/src/exports_info.rs

```diff
@@ -1829,3 +1829,247 @@ macro_rules! debug_exports_info {
+    Arc::new(
+      self
+        .inner
+        .lock()
+        .map(|mut mg| {
+          mg.get_export_info_mut_by_id(export_info_id)
+            .get_max_target()
+            .to_owned()
+        })
+        .expect("Can get export info id"),
+    )
+  }
+
+  fn run_resolve_filter(
+    &self,
+    target: &ResolvedExportInfoTarget,
+    resolve_filter: &ResolveFilterFnTy,
+  ) -> bool {
+    self
+      .inner
+      .lock()
+      .map(|mg| resolve_filter(target, &mg))
+      .expect("Can run resolve filter")
+  }
+
+  fn get_export_info(&self, export_info_id: &ExportInfoId) -> Arc<ExportInfo> {
+    Arc::new(
+      self
+        .inner
+        .lock()
+        .map(|mg| mg.get_export_info_by_id(export_info_id).to_owned())
+        .expect("Can get export info"),
+    )
+  }
+
+  fn get_read_only_export_info(
+    &self,
+    name: &Atom,
+    module_identifier: &ModuleIdentifier,
+  ) -> Option<Arc<ExportInfo>> {
+    self
+      .inner
+      .lock()
+      .map(|mg| {
+        mg.get_read_only_export_info(module_identifier, name.to_owned())
+          .map(|i| Arc::new(i.to_owned()))
+      })
+      .expect("Can get read only export info")
+  }
+
+  fn get_module_meta_exports_type(
+    &self,
+    module_identifier: &ModuleIdentifier,
+  ) -> Option<Arc<BuildMetaExportsType>> {
+    self
+      .inner
+      .lock()
+      .map(|mg: std::sync::MutexGuard<'_, &mut ModuleGraph>| {
+        mg.module_by_identifier(module_identifier)
+          .and_then(|m| m.build_meta())
+          .map(|meta| meta.exports_type)
+          .map(Arc::new)
+      })
+      .expect("Can get read only export info")
+  }
+}
+
+pub struct ImmutableModuleGraph<'a> {
+  inner: &'a ModuleGraph,
+}
+
+impl<'a> ImmutableModuleGraph<'a> {
+  pub fn new(mg: &'a ModuleGraph) -> Self {
+    Self { inner: mg }
+  }
+}
+
+impl<'a> ModuleGraphAccessor<'a> for ImmutableModuleGraph<'a> {
+  fn get_export_info_id(
+    &self,
+    name: &Atom,
+    module_identifier: &ModuleIdentifier,
+  ) -> Arc<ExportInfoId> {
+    Arc::new(
+      self
+        .inner
+        .module_graph_module_by_identifier(module_identifier)
+        .expect("should have mgm")
+        .exports
+        .get_read_only_export_info(name, self.inner)
+        .id,
+    )
+  }
+
+  fn get_max_target(
+    &self,
+    export_info_id: &ExportInfoId,
+  ) -> Arc<HashMap<Option<DependencyId>, ExportInfoTargetValue>> {
+    Arc::new({
+      let export_info_id = self.inner.get_export_info_by_id(export_info_id);
+
+      if export_info_id.max_target_is_set {
+        export_info_id.get_max_target_readonly().to_owned()
+      } else {
+        // FIXME:
+        // max target should be cached
+        // but when moduleGraph is immutable and no cached max target
+        // generate a new one, this should be removed
+        // when module graph in get_exports and get_referenced_exports
+        // is refactored to mutable
+        export_info_id._get_max_target()
```

Comment:
Cause module.get_exports_type() is called in get_exports and get_referenced_exports which is called in finishModules hook and optimizeDependencies hook by flag dependency plugins. And also some methods in depdencies are reused in code generation, so this need another refactor